lib/puppet_x/puppetlabs/cd4pe_job_helper.rb

```diff
@@ -0,0 +1,166 @@
+require 'json'
+
+module PuppetX::Puppetlabs
+  # A class to help with the job creation / running process for CD4PE
+  class CD4PEJobHelper < Object
+    attr_reader :docker_run_args
+
+    MANIFEST_TYPE = { 
+      :JOB => "JOB", 
+      :AFTER_JOB_SUCCESS => "AFTER_JOB_SUCCESS", 
+      :AFTER_JOB_FAILURE => "AFTER_JOB_FAILURE" }
+
+    def initialize(working_dir:, docker_image: nil, docker_run_args: nil)
+      @docker_image = docker_image
+      @docker_run_args = docker_run_args
+      @docker_based_job = !docker_image.nil?
+      @working_dir = working_dir
+
+      @local_jobs_dir = File.join(@working_dir, "cd4pe_job", "jobs", "unix")
+      @local_repo_dir = File.join(@working_dir, "cd4pe_job", "repo")
+      
+      @docker_run_args = docker_run_args.nil? ? '' : docker_run_args.join(' ')
+    end
+
+    def set_job_env_vars(task_params)
+      ENV["WEB_UI_ENDPOINT"] = task_params['cd4pe_web_ui_endpoint']
+      ENV["JOB_TOKEN"] = task_params['cd4pe_token']
+      ENV["JOB_OWNER"] = task_params['cd4pe_job_owner']
+      ENV["JOB_INSTANCE_ID"] = task_params['job_instance_id']
+    
+      user_specified_env_vars = task_params['env_vars']
+      if (!user_specified_env_vars.nil?)
+        user_specified_env_vars.each do |var|
+          pair = var.split("=")
+          key = pair[0]
+          value = pair[1]
+          ENV[key] = value
+      end
+    end
+
+    def make_working_dir(working_dir)
+      Dir.mkdir(working_dir) unless File.exists?(working_dir)
+    end
+
+    def get_job_script_and_control_repo
+      target_file = File.join(@working_dir, "cd4pe_job.tar.gz")
+      client = PuppetX::Puppetlabs::CD4PEClient.new
+      response = client.get_job_script_and_control_repo(target_file)
+      case response
+      when Net::HTTPNotFound
+        raise "Message: #{response.body}\nCode: #{response.code}"
+      when Net::HTTPServerError
+        raise "Unknown HTTP Error with code: #{response.code} and body #{response.body}"
+      end
+
+      target_file
+    end
+
+    def run_job
+      result = execute_manifest(MANIFEST_TYPE[:JOB])
+      
+      if (result[:exit_code] == 0)
+        on_job_success(result)
+      else
+        on_job_failure(result)
+      end
+    end
+
+    def on_job_success(result)
+      output = {}
+      output[:job] = {
+        exit_code: result[:exit_code], 
+        message: result[:message]
+      }
+    
+      # if a AFTER_JOB_SUCCESS script exists, run it now!
+      run_after_success = File.exists?(File.join(@local_jobs_dir, MANIFEST_TYPE[:AFTER_JOB_SUCCESS]))
+      if (run_after_success)
+        @@logs.push("#{MANIFEST_TYPE[:AFTER_JOB_SUCCESS]} script specified.")
+        success_script_result = execute_manifest(MANIFEST_TYPE[:AFTER_JOB_SUCCESS])
+        output[:after_job_success] = {
+          exit_code: success_script_result[:exit_code], 
+          message: success_script_result[:message]
+        }
+      end
+    
+      output
+    end
+
+    def on_job_failure(result)
```

Comment:
Seems like `on_job_failure` and `on_job_success` are identical enough that they could be one method, with internal if checks on the `result[:exit_code]` where behavior is different.

This code might flow better if we didn't do the `if exit code == 0` right away, but instead just passed around the `result` object and let the exit code checks happen inside the method, at a smaller scope.

Just some design thought, not a change we need to make.

Reply:
originally they were not quite as similar, but they def are now. Updating.